### PR TITLE
fix(DTFS2-6963): Holistic user service approach to block users who have `disabled` `true`

### PR DESCRIPTION
**Pull request by ChristianMcc**
## Introduction
We needed an approach we can use everywhere which validates the status of a user and if they are disabled

## Resolution
Introduce a user service to take care of validating a user's status and disabled state. Add this on the relevant login endpoints during the 'login in progress' part of the 2fa section. Then after this add it to the passport validation so that we now check on every API request if the user is in a valid state, meaning any disabled users don't have access to the system straight away (rather than having a full 12 hour session to do whatever they want with!)

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ run-name: Executing test QA on ${{ github.repository }} 🚀
 
 on:
   pull_request:
-    branches: [main, fix/DTFS2-6910/fix-issue-withh-login-page-redirect]
+    branches: [main, fix/DTFS2-6910/fix-issue-withh-login-page-redirect, fix/DTFS2-6963/patch-user-disabled-feature-to-sign-in-link]
```
Comment (AlexBramhill) [inline .github/workflows/test.yml:11]: reminder to revert

File: portal-api/src/v1/users/login.controller.js [hunk 2]
```diff
@@ -1,9 +1,8 @@
-const { STATUS } = require('../../constants/user');
 const utils = require('../../crypto/utils');
-const { userIsBlocked, userIsDisabled, usernameOrPasswordIncorrect } = require('../../constants/login-results');
+const { usernameOrPasswordIncorrect } = require('../../constants/login-results');
 const { findByUsername, incrementFailedLoginCount, updateSessionIdentifier } = require('./controller');
 
-module.exports.login = (username, password) =>
+module.exports.login = (username, password, userService) =>
```
Comment (AlexBramhill) [inline portal-api/src/v1/users/login.controller.js:5]: We decided on this? Happy to have it

File: portal-api/src/v1/users/passport.js [hunk 3]
```diff
@@ -54,13 +64,21 @@ const baseAuthenticationConfiguration = ({ name, passport, additionalValidation,
...
-const loginCompleteAuth = (passport) => {
-  const additionalValidation = (jwtPayload) => jwtPayload.loginStatus === LOGIN_STATUSES.VALID_2FA;
-  baseAuthenticationConfiguration({ name: 'login-complete', passport, additionalValidation });
+const loginCompleteAuth = (passport, userService) => {
+  const name = 'login-complete';
+  const additionalPayloadValidation = (jwtPayload) => jwtPayload.loginStatus === LOGIN_STATUSES.VALID_2FA;
+  const additionalUserValidation = (user, jwtPayload) => {
+    if (userService.isUserBlockedOrDisabled(user)) {
+      console.error("User with username %s is blocked or disabled for '%s' strategy", jwtPayload.username, name);
+      return true;
```
Comment (AlexBramhill) [inline portal-api/src/v1/users/passport.js:73]: Surprising -- Could we add in a VALIDATION_RESULT const that is either Passed or Failed

**Comment by ChristianMcc** [comment]
I think I'd like to add an E2E test which has a user logged in with a session, who then gets disabled, and test that we are routed to the login page and session is destroyed. I think Oscar added a nice cypress command that can help with that so will get to it in the morning!

**Comment by AlexBramhill** [comment]
>E2E test which has a user logged in with a session, who then gets disabled, and test that we are routed to the login page and session is destroyed 

We might already have this! :) I think I found oscar's current test in e2e does this (logs in, disables user, clicks on link).

File: e2e-tests/portal/cypress/e2e/journeys/login/navigating-using-sign-in-link.spec.js [hunk 4]
```diff
@@ -140,6 +140,21 @@ context('navigating using sign in link', () => {
       signInLink.visit({ token: NOT_EXPIRED_SIGN_IN_TOKEN.signInTokenFromLink, userId: bank1Maker1Id }, { failOnStatusCode: false });
       signInLink.shouldDisplayAccountSuspended();
     });
+
+    it('Opening a valid sign in link and logging in but then disabling the user and visiting the landing page will redirect the user to the login page', () => {
```
Comment (ChristianMcc) [inline e2e-tests/portal/cypress/e2e/journeys/login/navigating-using-sign-in-link.spec.js:144]: this tests disabling a user once they are logged in happily and making sure they get logged out

File: portal-api/api-tests/v1/users/login-with-sign-in-link.api-test.js [hunk 5]
```diff
@@ -300,15 +300,15 @@ describe('POST /users/:userId/sign-in-link/:signInToken/login', () => {
             });
           });
 
-          it('returns a user blocked 403 error', async () => {
+          it('returns a user disabled 403 error', async () => {
```
Comment (ChristianMcc) [inline portal-api/api-tests/v1/users/login-with-sign-in-link.api-test.js:303]: I updated the API to send the correct disabled or blocked error for the user, the frontend as per initial requirements still shows the blocked page for either a disabled or blocked user

File: portal-api/src/v1/users/controller.test.js [hunk 6]
```diff
@@ -48,4 +48,32 @@ describe('user controller', () => {
       expect(mockCallback).toHaveBeenCalledTimes(1);
     });
   });
+
+  describe('createPasswordToken', () => {
```
Comment (ChristianMcc) [inline portal-api/src/v1/users/controller.test.js:52]: added tests to check when trying to reset password that the user isn't blocked or disabled, if they are then we don't return a token

File: portal-api/src/v1/users/passport.js [hunk 7]
```diff
@@ -39,28 +49,45 @@ const baseAuthenticationConfiguration = ({ name, passport, additionalValidation,
...
-          : user && user.sessionIdentifier === jwtPayload.sessionIdentifier;
+        if (sessionIdentifierValidation(user, jwtPayload) === PASSPORT_VALIDATION_RESULTS.FAILED) {
+          console.error("Failed JWT validation for '%s' strategy", name);
+          return done(null, false);
+        }
 
-        if (validation) {
-          const additionalFields = getAdditionalReturnedFields(user);
-          return done(null, { ...sanitize(user), ...additionalFields });
+        if (additionalUserValidation && additionalUserValidation(user, jwtPayload) === PASSPORT_VALIDATION_RESULTS.FAILED) {
```
Comment (ChristianMcc) [inline portal-api/src/v1/users/passport.js:57]: updated validation in passport so that if we provide extra user validation we provide a clear console log and don't authenticate if they are blocked or disabled once they are on the login-complete passport flow

**Review by avaitonis — DISMISSED** [review]
LGTM

File: portal-api/src/v1/errors/user-disabled.error.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class UserDisabledError extends Error {
```
Comment (abhi-markan) [inline portal-api/src/v1/errors/user-disabled.error.js:1]: Can we please add some documentation on the class, appreciate other classes do not have it either.
Happy to be a separate PR if required.

File: portal-api/src/v1/errors/user-disabled.error.test.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+const UserDisabledError = require('./user-disabled.error');
+
+describe('UserDisabledError', () => {
+  const userId = 'exampleUserId';
```
Comment (abhi-markan) [inline portal-api/src/v1/errors/user-disabled.error.test.js:4]: Will there ever be a scenario, where the `userId` will be empty?

File: portal-api/src/v1/users/controller.test.js [hunk 10]
```diff
@@ -48,4 +48,32 @@ describe('user controller', () => {
...
+    let userService;
+
+    it('if the user is blocked or disabled then no token is returned', async () => {
+      const mockFindOne = jest.fn().mockResolvedValue(TEST_USER);
+      when(db.getCollection).calledWith('users').mockResolvedValue({ findOne: mockFindOne });
+      userService = {
+        isUserBlockedOrDisabled: jest.fn().mockImplementationOnce(() => true),
+      };
+
+      const token = await createPasswordToken('mockEmail', userService);
```
Comment (abhi-markan) [inline portal-api/src/v1/users/controller.test.js:62]: `mockEmail` from a pre-defined constant please.

File: portal-api/src/v1/users/login.controller.js [hunk 11]
```diff
@@ -21,12 +22,16 @@ module.exports.login = (username, password) =>
...
-      if (user.disabled) {
-        return resolve({ error: userIsDisabled });
-      }
-
-      if (user['user-status'] === STATUS.BLOCKED) {
-        return resolve({ error: userIsBlocked });
+      try {
+        userService.validateUserIsActiveAndNotDisabled(user);
+      } catch (e) {
+        console.error(e);
```
Comment (abhi-markan) [inline portal-api/src/v1/users/login.controller.js:28]: Please standardise `console.error`.

File: portal/server/routes/middleware/validatePartialAuthToken/index.test.js [hunk 12]
```diff
@@ -62,15 +61,32 @@ describe('validatePartialAuthToken', () => {
...
-          .mockRejectedValueOnce(new Error('test error'));
+        when(api.validatePartialAuthToken).calledWith(userToken).mockRejectedValueOnce(new Error('test error'));
       });
 
       itDoesNotCallNext();
       itDestroysTheSession();
       itRedirectsToLoginPage();
     });
+
+    describe('when portal api successfully validates the token as a login complete', () => {
```
Comment (abhi-markan) [inline portal/server/routes/middleware/validatePartialAuthToken/index.test.js:72]: Should we add assertions where we check following 2FA specific properties values:

* signInLinkSendCount
* signInToken

File: portal-api/src/v1/users/user.service.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
+const { STATUS } = require('../../constants/user');
+const UserBlockedError = require('../errors/user-blocked.error');
+const UserDisabledError = require('../errors/user-disabled.error');
+
+class UserService {
```
Comment (abhi-markan) [inline portal-api/src/v1/users/user.service.js:5]: Please add some documentation.

File: portal-api/src/v1/users/user.service.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+      throw new UserDisabledError(user._id);
+    }
+
+    if (isUserBlocked) {
+      throw new UserBlockedError(user._id);
+    }
+  }
+
+  isUserBlockedOrDisabled(user) {
+    return user['user-status'] === STATUS.BLOCKED || user.disabled === true;
```
Comment (abhi-markan) [inline portal-api/src/v1/users/user.service.js:20]: Not all the users will have `disabled` property (deprecated property) can this please be converted to optional chaining `user?.disabled`

Comment (abhi-markan) [reply portal-api/src/v1/users/user.service.js:20]: Missing function documentation.

File: portal-api/src/v1/users/user.service.test.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+  describe('when the user is blocked', () => {
+    let blockedUser;
+    beforeEach(() => {
+      blockedUser = produce(testUser, (draft) => {
+        draft['user-status'] = STATUS.BLOCKED;
+      });
+    });
+
+    it('validating user throws a UserBlockedError', async () => {
```
Comment (abhi-markan) [inline portal-api/src/v1/users/user.service.test.js:27]: Do we need an assertion where `disabled` property does not exists and `user-status: active`?

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Good work, few comments to address please.

File: portal-api/src/v1/errors/user-disabled.error.test.js (hunk 9 above)
Comment (ChristianMcc) [reply portal-api/src/v1/errors/user-disabled.error.test.js:4]: Nope, we shouldn't ever throw this error as it references a user being blocked (and we need the userId to know the user)

File: portal-api/src/v1/users/controller.test.js (hunk 6 above)
Comment (abhi-markan) [reply portal-api/src/v1/users/controller.test.js:54]: Test name makes sense but any textual explanation is always helpful when being visited in fresh.

File: portal-api/src/v1/users/login.controller.js [hunk 16]
```diff
@@ -21,12 +22,16 @@ module.exports.login = (username, password) =>
...
+        if (e instanceof UserBlockedError) {
+          return resolve({ error: userIsBlocked });
+        }
+        if (e instanceof UserDisabledError) {
+          return resolve({ error: userIsDisabled });
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply portal-api/src/v1/users/login.controller.js:33]: Makes sense. Thanks

File: portal-api/src/v1/users/user.service.js (hunk 14 above)
Comment (abhi-markan) [reply portal-api/src/v1/users/user.service.js:30]: Thanks

**Review by abhi-markan — APPROVED** [review]

**Review by abhi-markan — APPROVED** [review]
